File: treinamento/ECNA23/e.py


def dfs(pos, val, resp):
    if pos == -1:
        resp = min(resp, val)
        return resp

    if s[pos] == 'O':
        if (val-1)%3 == 0 and ((val-1)/3)%2 == 1:
            if (val-1)/3 <= 1:
                return 10000000000000000000
            return dfs(pos-1, (val-1)/3, resp)

    elif s[pos] == 'E':
        if int((val*2)).bit_count() == 1:
            return 10000000000000000000
        return dfs(pos-1, val*2, resp)

    return 10000000000000000000

s = input()

for i in range(len(s)-1):
    if s[i] == 'O' and s[i+1] == 'O':
        print("INVALID")
        exit(0)

if s[len(s)-1] != 'O':
    print("INVALID")
    exit(0)

aux = 1
resp = 10000000000000000000

for i in range(300):
    resp = min(resp,dfs(len(s)-1, aux, resp))
    aux*=2

if resp < 10000000000000000000:
    print(int(resp))
else:
    print("INVALID")

File: treinamento/ECNA23/e.cpp
#include <bits/stdc++.h>
using namespace std;
typedef unsigned long long ll;

string s;
ll resp = 1e17;

void dfs(int pos, ll val){
    // cout << pos << " " <<  val << endl;
    if (val < 0) return;
    if (pos == -1){
        resp = min(resp, val);
        return;
        // cout << val << endl;
    }
    if (s[pos] == 'O'){
        if ((val-1)%3 == 0 and ((val-1)/3)%2 == 1){
            if ((val-1)/3 <= 1) return;
            dfs(pos-1, (val-1)/3);
        }
    }
    else if (s[pos] == 'E'){
        if (__builtin_popcountll(val*2) == 1) return;
        dfs(pos-1, val*2);
    }

}


int main(){
    cin >> s;

    for (int i=0; i<(int)s.size()-1; i++){
        if (s[i] == 'O' and s[i+1] == 'O'){
            cout << "INVALID\n";
            return 0;
        }
        if (s[i] != 'E' and s[i] != 'O'){
            cout << "INVALID\n";
            return 0;
        }
    }
    if (s.back() != 'O'){
        cout << "INVALID\n";
        return 0;
    }
    for (ll i = 0; i<=70; i++){
        ll aux = 1;
        dfs((int)s.size()-1,(aux << i));
        // cout << (aux << i) << endl;
        //cout << i << " " << resp << endl;
    }
    if (resp < 1e17){
        cout << resp << endl;
    }

    else cout << "INVALID\n";
    

}